arraycontext/container/traversal.py

```diff
@@ -966,14 +965,21 @@ def treat_as_scalar(x: Any) -> bool:
             return (
                 not isinstance(x, np.ndarray)
                 # This condition is whether "ndarrays should broadcast inside x".
-                and np.ndarray not in x.__class__._outer_bcast_types)
+                and NumpyObjectArray not in x.__class__._outer_bcast_types)
+
+    a_is_ndarray = isinstance(a, np.ndarray)
+    b_is_ndarray = isinstance(b, np.ndarray)
+
+    if a_is_ndarray and a.dtype != object:
+        raise TypeError("passing a non-object numpy array is not allowed")
+    if b_is_ndarray and b.dtype != object:
+        raise TypeError("passing a non-object numpy array is not allowed")
```

Comment:
@majosm This bit is new. I don't recall why it was important to treat non-object-arrays as scalars, but it seems like a weird inconsistency. Unless you strongly object, I'm hoping to remove that.